File: sdk/javascript/examples/query-noi.ts
/**
 * Example: Query NOI snapshots for a crop cycle with filters.
 *
 * Usage:
 *   npx tsx examples/query-noi.ts
 */

import { KokonutClient } from '../src/index.js';

const client = new KokonutClient('http://localhost:8055', {
  token: process.env.KOKONUT_TOKEN,
});

async function main() {
  await client.login('admin@example.com', 'password123');

  const cropCycles = await client.cropCycles.list({
    filter: { status: { _in: ['harvested', 'completed'] } },
    sort: ['-planting_date'],
    limit: 5,
  });

  console.log(`Found ${cropCycles.length} completed crop cycles\n`);

  for (const cycle of cropCycles) {
    console.log(`--- ${cycle.cycle_name} (${cycle.id}) ---`);
    console.log(`  Planted: ${cycle.planting_date}`);
    console.log(`  Harvested: ${cycle.actual_harvest_date ?? 'N/A'}`);
    console.log(`  Yield: ${cycle.yield_actual ?? 'N/A'} ${cycle.yield_unit ?? ''}`);

    const harvests = await client.harvestEvents.listByCropCycle(cycle.id, {
      sort: ['-harvest_date'],
    });
    console.log(`  Harvest events: ${harvests.length}`);

    const sales = await client.salesEvents.listByCropCycle(cycle.id, {
      sort: ['-sale_date'],
    });
    const totalRevenue = sales.reduce((sum: number, s: any) => sum + (s.total_amount ?? 0), 0);
    console.log(`  Sales events: ${sales.length}`);
    console.log(`  Total revenue: $${totalRevenue.toLocaleString()}`);

    const expenses = await client.expenseEvents.listByCropCycle(cycle.id);
    const totalExpenses = expenses.reduce((sum: number, e: any) => sum + (e.amount ?? 0), 0);
    console.log(`  Expense events: ${expenses.length}`);
    console.log(`  Total expenses: $${totalExpenses.toLocaleString()}`);

    const noi = totalRevenue - totalExpenses;
    const margin = totalRevenue > 0 ? ((noi / totalRevenue) * 100).toFixed(1) : '0.0';
    console.log(`  NOI: $${noi.toLocaleString()} (${margin}% margin)\n`);
  }

  const reports = await client.reports.listByType('crop_noi', {
    sort: ['-created_at'],
    limit: 5,
  });
  console.log(`\nRecent NOI report snapshots: ${reports.length}`);
  for (const report of reports) {
    console.log(`  ${report.id} - ${report.created_at}`);
    console.log(`    Hash: ${report.snapshot_hash}`);
  }
}

main().catch(console.error);

File: sdk/python/examples/query_noi.py
"""Example: Query NOI snapshots for completed crop cycles.

Usage:
    KOKONUT_TOKEN=... python examples/query_noi.py
"""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from kokonut import KokonutClient, ListOptions

BASE_URL = os.environ.get("KOKONUT_BASE_URL", "http://localhost:8055")
TOKEN = os.environ.get("KOKONUT_TOKEN")


def main() -> None:
    client = KokonutClient(BASE_URL, token=TOKEN)

    if not TOKEN:
        client.login("admin@example.com", os.environ.get("ADMIN_PASSWORD", "changeme"))

    crop_cycles = client.crop_cycles.list(
        ListOptions(
            filter={"status": {"_in": ["harvested", "completed"]}},
            sort=["-planting_date"],
            limit=5,
        )
    )

    print(f"Found {len(crop_cycles)} completed crop cycles\n")

    for cycle in crop_cycles:
        print(f"--- {cycle.get('cycle_name', cycle['id'])} ---")
        print(f"  Planted:    {cycle.get('planting_date', 'N/A')}")
        print(f"  Harvested:  {cycle.get('actual_harvest_date', 'N/A')}")
        print(f"  Yield:      {cycle.get('actual_yield', 'N/A')} {cycle.get('actual_yield_unit', '')}")

        harvests = client.harvest_events.list_by_crop_cycle(
            cycle["id"], ListOptions(sort=["-harvest_date"])
        )
        print(f"  Harvests:   {len(harvests)}")

        sales = client.sales_events.list_by_crop_cycle(
            cycle["id"], ListOptions(sort=["-sale_date"])
        )
        total_revenue = sum(s.get("total_amount", 0) or 0 for s in sales)
        print(f"  Sales:      {len(sales)}")
        print(f"  Revenue:    ${total_revenue:,.2f}")

        expenses = client.expense_events.list_by_crop_cycle(cycle["id"])
        total_expenses = sum(e.get("amount", 0) or 0 for e in expenses)
        print(f"  Expenses:   {len(expenses)}")
        print(f"  Costs:      ${total_expenses:,.2f}")

        noi = total_revenue - total_expenses
        margin = (noi / total_revenue * 100) if total_revenue > 0 else 0
        print(f"  NOI:        ${noi:,.2f} ({margin:.1f}% margin)\n")

    # Also query stored NOI snapshots
    reports = client.reports.list_by_type(
        "crop_noi", ListOptions(sort=["-created_at"], limit=5)
    )
    print(f"Recent NOI report snapshots: {len(reports)}")
    for report in reports:
        print(f"  {report['id']} - {report.get('created_at', 'N/A')}")
        if report.get("snapshot_hash"):
            print(f"    Hash: {report['snapshot_hash']}")


if __name__ == "__main__":
    main()
